### PR TITLE
Solutions and bookings

### PR DESCRIPTION
**Pull request by buhaytza2005**
I've added support for the Bookings API. It includes businesses, customers, appointments and custom questions. The PR includes a couple of examples too.

### CODE REVIEW

**Comment by sreeise** [comment]
Hey, thanks for doing this and apologies for not getting to this sooner. I have been very busy as of late.

I am currently in the middle of releasing the next version and once that is done I will take a look at this
within the next couple of days.

File: src/solutions/appointments/request.rs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        body: true
+    );
+    get!(
+        doc: "Get appointments from solutions",
+        name: list_appointments,
+        path: "/appointments"
+    );
+    get!(
+        doc: "Get the number of the resource",
+        name: appointments,
```
Comment (sreeise) [inline src/solutions/appointments/request.rs:25]: Update this to `get_appointments_count` by updating codgen here: https://github.com/sreeise/graph-rs-sdk/blob/b35f98e641ed17ad13f21c5729d1ffd8894394cc/graph-codegen/src/settings/method_macro_modifier.rs#L55

File: src/solutions/custom_questions/request.rs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        body: true
+    );
+    get!(
+        doc: "Get customQuestions from solutions",
+        name: list_custom_questions,
+        path: "/customQuestions"
+    );
+    get!(
+        doc: "Get the number of the resource",
+        name: custom_questions,
```
Comment (sreeise) [inline src/solutions/custom_questions/request.rs:25]: Update count method name to `get_custom_questions_count` through codegen.

File: src/solutions/customers/request.rs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        body: true
+    );
+    get!(
+        doc: "Get customers from solutions",
+        name: list_customers,
+        path: "/customers"
+    );
+    get!(
+        doc: "Get the number of the resource",
+        name: customers,
```
Comment (sreeise) [inline src/solutions/customers/request.rs:25]: Same as other count methods. Change to `get_customers_count`

File: src/solutions/services/request.rs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        body: true
+    );
+    get!(
+        doc: "Get services from solutions",
+        name: list_services,
+        path: "/services"
+    );
+    get!(
+        doc: "Get the number of the resource",
+        name: services,
```
Comment (sreeise) [inline src/solutions/services/request.rs:25]: Update to `get_services_count`

File: examples/solutions/business.rs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+    let body = Body::from(data.to_string());
+    
+    let resp = client
+        .solutions()
+        .booking_businesses()
+        .create_booking_businesses(body)
+        .send()
+        .await
+        ;
```
Comment (sreeise) [inline examples/solutions/business.rs:27]: semicolon should be at the end of .await

File: src/solutions/booking_businesses/request.rs [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        body: true
+    );
+    get!(
+        doc: "Get bookingBusinesses from solutions",
+        name: list_booking_businesses,
+        path: "/bookingBusinesses"
+    );
+    get!(
+        doc: "Get the number of the resource",
+        name: booking_businesses,
```
Comment (sreeise) [inline src/solutions/booking_businesses/request.rs:26]: Update this to get_booking_businesses_count by updating codgen here: 
https://github.com/sreeise/graph-rs-sdk/blob/b35f98e641ed17ad13f21c5729d1ffd8894394cc/graph-codegen/src/settings/method_macro_modifier.rs#L55

File: src/solutions/booking_businesses/request.rs [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        body: true,
+        params: booking_appointment_id
+    );
+    post!(
+        doc: "Invoke action cancel",
+        name: cancel,
+        path: "/bookingBusinesses/{{RID}}/calendarView/{{id}}/cancel",
+        body: true,
+        params: booking_appointment_id
+    );
```
Comment (sreeise) [inline src/solutions/booking_businesses/request.rs:100]: [Calendar view](https://github.com/sreeise/graph-rs-sdk/blob/329201426c15b36c3b685d50ae15c59ee1753694/src/users/calendar_view/request.rs) has its own code generation under the users folder. We should use a link to this client as it has all the same methods.

You will need to add calendarView to the filterPath call in codegen for BookingBusinessesIdApiClient and add an import for calendarView and then add an ApiClientLink in codgen for linking to calendarView as a method call.

File: src/solutions/staff_members/request.rs [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+        body: true
+    );
+    get!(
+        doc: "Get staffMembers from solutions",
+        name: list_staff_members,
+        path: "/staffMembers"
+    );
+    get!(
+        doc: "Get the number of the resource",
+        name: staff_members,
```
Comment (sreeise) [inline src/solutions/staff_members/request.rs:25]: Update to `get_staff_members_count`

**Review by sreeise — CHANGES_REQUESTED** [review]
This is really good and I think your probably the first person, besides me of course, to add the necessary settings and what not in order to generate a client correctly. So thanks for taking the time to understand it.

There are a few things that need to be updated but they are minor as far as changes go.

Can you also make sure to run `cargo fmt` so everything is formatted.

**Comment by buhaytza2005** [comment]
Hi @sreeise! Thanks for taking the time to review the PR.

Re the renaming of the methods https://github.com/sreeise/graph-rs-sdk/pull/471/commits/f69985a224183e515054cddbb0a82725c450beca contains the changes but when I regenerated the code the methods were not replaced/renamed. Am I missing a trick or is the code incorrect?

I assumed that:

```rust
MethodMacroModifier::fn_name_and_path(
			"appointments", "/appointments/$count",
			GeneratedMacroType::FnName("get_appointments_count")
		),
```
would be the correct implementation, am I wrong in my assumption? Or is there anything else that I would need to add?

Thanks,
Mike

**Comment by sreeise** [comment]
> Hi @sreeise! Thanks for taking the time to review the PR.
> 
> Re the renaming of the methods: [f69985a](https://github.com/sreeise/graph-rs-sdk/commit/f69985a224183e515054cddbb0a82725c450beca) contains the changes but when I regenerated the code the methods were not replaced/renamed. Am I missing a trick or is the code incorrect?
> 
> I assumed that:
> 
> ```rust
> MethodMacroModifier::fn_name_and_path(
> 			"appointments", "/appointments/$count",
> 			GeneratedMacroType::FnName("get_appointments_count")
> 		),
> ```
> 
> would be the correct implementation, am I wrong in my assumption? Or is there anything else that I would need to add?
> 
> Thanks, Mike

You have the methods matching ResourceIdentity::Solutions but that won't work here. You'll have to add a match for each of the ResourceIdentity enums that match the api client struct where the method is implemented for each.

For instance `ResourceIdentity::BookingBusinesses`:

```rust
ResourceIdentity::BookingBusinesses => vec![
  MethodMacroModifier::fn_name_and_path(
	"booking_businesses", "/bookingBusinesses/$count",
	GeneratedMacroType::FnName("get_booking_businesses_count")
  ),
]
```

File: src/solutions/booking_businesses/request.rs (hunk 7 above)
Comment (buhaytza2005) [reply src/solutions/booking_businesses/request.rs]: That's been done. Got caught out again with adding the link-ups in the wrong match but should all be good now :)

**Comment by sreeise** [comment]
This all looks correct. There is only one issue that this is actually my fault.

I forgot that in 2.0 the macro used for the struct definitions of clients was changed from `resource_api_client` to `api_client` but I guess I forgot to update 1 part of codegen where the name is formatted and written out:

https://github.com/sreeise/graph-rs-sdk/blob/b35f98e641ed17ad13f21c5729d1ffd8894394cc/graph-codegen/src/macros/macro_queue_writer.rs#L451

That is my fault I apologize.

You can do one of two things:

Either manually change resource_api_client calls in the solutions directory or update that codegen line above and rerun the codegen. Obviously that will have to get updated at some point if you just change your solutions directory but I don't want to require you to fix my mistake unless your willing and want to.

**Comment by buhaytza2005** [comment]
Happy to fix, love helping.

I saw that after merge with 2.0 something went wrong, but thought I messed something up from my end 🤣 .
I'll update the PR and resubmit sometime this weekend.

**Comment by buhaytza2005** [comment]
All done but I had to manually replace `resource_api_client` due to #478 as I can't regenerate the code.

**Comment by sreeise** [comment]
> All done but I had to manually replace `resource_api_client` due to #478 as I can't regenerate the code.

Ah gotcha. I'll take a look at that here soon and hopefully get it fixed. Thanks for doing this and good work. Its much appreciated.

**Review by sreeise — APPROVED** [review]

**Pull request merged**